### PR TITLE
[DRAFT] [lworld] Test scalarization and call conventions for different value class shapes

### PR DESCRIPTION
**Pull request by galderz**
This is a **very early draft** for a template-based test (temporarily called `TestOne`) that addresses some of the requirements of JDK-8364421.

**NOTE**: the PR also includes the commits to add template-based testing framework since that's not yet in `lworld`, and also adds https://github.com/openjdk/valhalla/commit/68e4a60c1ff62f4d0b8b31cb0bd729f1d67748e2 so that Boolean types are supported in `checkEQ`.

This test has surfaced [JDK-8365590](https://bugs.openjdk.org/browse/JDK-8365590) bug, and as I focus on that bug, I thought it'd be a good moment to share what I've done so far with the test and get some feedback.

The test is very simple so far, creating value classes that contain a single field, testing for each of the primitive types. I also started testing having this field be a primitive array type and that's when I encountered [JDK-8365590](https://bugs.openjdk.org/browse/JDK-8365590).

I've tried to create value classes that had multiple fields, but I've hit some obstacles here so I've just added a very basic test:

First, I tried to use data names to create value classes that would have N fields, and have these fields initialized via a constructor, but I couldn't get it to work. So the example present in the test uses external data structures to keep track of these fields, their types, their initial values...etc. I have discussed these issues in more detail with @chhagedorn and bounced some ideas of how to improve data names, but these need to be discussed in more detail.

The second obstacle I found is related to nuances on when value classes with multiple fields will be scalarized, when these classes are part of another value classes. Depending on the combination of fields, they can be considered that they can be flattened or not, so handling this in a test like this would need to track the sizes of each fields, add them up, and decide whether the test should be positive on scalarization or negative. I traced this logic to `FieldLayoutBuilder::compute_inline_class_layout` and the way `_nullable_layout_size_in_bytes` is then used to decide whether to flatten or not. I've been using `PrintFieldLayout` option to see what nullable flat layout values each of these box instances would have.

So, given these complexities of multiple fields, I'm might keep `TestOne` (note: name is temporary) focused on value classes with a single field. And then have another test with multiple fields.

### CODE REVIEW

**Comment by eme64** [comment]
And it would be nice to see if we cannot use `DataName` for fields, and `StructuralName` for methods (maybe it is just a small issue with how to use the framework). But it's not a necessity: if you feel like modeling it separately that is also fine.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/templating/TestOne.java [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+    static TemplateToken hashMethod(List<FieldConstant> fields) {
+        return Template.make(() -> body(
+            """
+            int hash() {
+                return
+            """,
+            fields.stream().map(TestOne::hashField).toList(),
+            """
+                0;
+            }
```
[the remark this answers is not in the document]
Comment (galderz) [reply test/hotspot/jtreg/compiler/valhalla/inlinetypes/templating/TestOne.java:210]: True. I had in my mind that I needed to return a List of `TemplateToken` but I guess I can just convert those to string and join then for easier to read code.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/templating/TestOne.java [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        return type instanceof PrimitiveType pt
+            ? pt.boxedTypeName()
+            : type.name();
+    }
+
+    static TemplateToken multiFieldType(List<PrimitiveType> fieldTypes) {
+        final AtomicInteger fieldId = new AtomicInteger();
+        final List<FieldConstant> fields = fieldTypes.stream()
+            .map(fieldType -> FieldConstant.of(fieldId.getAndIncrement(), fieldType))
+            .toList();
```
[the remark this answers is not in the document]
Comment (galderz) [reply test/hotspot/jtreg/compiler/valhalla/inlinetypes/templating/TestOne.java:172]: `FieldConstant` is just quick way that I used to model what I couldn't model with `dataNames` in order to construct a `Box` type with 2 boolean fields.

The constants are leftover from a previous approach where I was trying to use them to both initialise a field with it, and then use it as the expected result verify that the test result was the constant that I had assigned in the field. Later on I found there are other ways to compute the expected result in ways that I don't have to track the value around, e.g. setting it to a static gold value or similar.

The id was just a way to number the fields that I added before I recognised I could use the `$` sign. I'm not sure if I still need it even though I'm not using dataNames for multi-field box types but I'll double check.

File: test/hotspot/jtreg/compiler/valhalla/inlinetypes/templating/TestOne.java [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+                // , "-XX:+PrintFieldLayout"
+                // , "-XX:+PrintInlining"
+            }}
+        );
+    }
+
+    record FieldConstant(Object value, int id, PrimitiveType type) {
+        String name() {
+            return "v" + id;
+        }
```
[the remark this answers is not in the document]
Comment (galderz) [reply test/hotspot/jtreg/compiler/valhalla/inlinetypes/templating/TestOne.java:104]: See my comment above

**Comment by galderz** [comment]
> And it would be nice to see if we cannot use DataName for fields, and StructuralName for methods (maybe it is just a small issue with how to use the framework). But it's not a necessity: if you feel like modeling it separately that is also fine.

I did try to use data names to fields and I have been able to use them fine for a single field box types, see `uniFieldTest`. I struggled to use data names for multi field box types, particularly when you have multiple fields of the same type, e.g. 2 boolean fields.

As for structural names, I did not explore them yet. I will look into them and see how they can help here.
